### PR TITLE
chore(codecs): Implement `JsonSerializer`

### PR DESCRIPTION
**Pull request by pablosichert**
Part of #8617, #10254 and towards the implementation of RFC #9864.

### CODE REVIEW

File: src/codecs/format/json.rs [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
 use chrono::Utc;
 use serde::{Deserialize, Serialize};
 use smallvec::{smallvec, SmallVec};
+use tokio_util::codec::Encoder;
 
 use crate::{
-    codecs::decoding::{BoxedDeserializer, Deserializer, DeserializerConfig},
+    codecs::{
+        decoding::{BoxedDeserializer, Deserializer, DeserializerConfig},
+        encoding::{BoxedSerializer, SerializerConfig},
```
Comment (spencergilbert) [inline src/codecs/format/json.rs:12]: Not being particularly familiar with the work so far... why isn't there a matching `encoding::Serializer` to the `decoding::Deserializer`?

**Review by spencergilbert — APPROVED** [review]

File: src/codecs/format/json.rs [hunk 2]
```diff
@@ -80,13 +84,60 @@ impl From<&JsonDeserializerConfig> for JsonDeserializer {
...
+}
+
+#[typetag::serde(name = "json")]
+impl SerializerConfig for JsonSerializerConfig {
+    fn build(&self) -> crate::Result<BoxedSerializer> {
+        Ok(Box::new(JsonSerializer))
+    }
+}
+
+/// Serializer that converts an `Event` to bytes by extracting the message key.
```
Comment (spencergilbert) [inline src/codecs/format/json.rs:105]: This comment seems incorrect? Maybe copied from `text`?

Comment (pablosichert) [reply src/codecs/format/json.rs]: Oopa, good catch 👀

File: src/codecs/format/json.rs (hunk 1 above)
Comment (pablosichert) [reply src/codecs/format/json.rs:12]: That's a good question! There is:

https://github.com/vectordotdev/vector/blob/eadf6c8a36296889898b54d6afd05d2cd3b10b6b/src/codecs/encoding.rs#L93-L97

it just happens to have a blanket implementation when we implement `tokio_util::codec::Encoder`:

https://github.com/vectordotdev/vector/blob/eadf6c8a36296889898b54d6afd05d2cd3b10b6b/src/codecs/encoding.rs#L99-L104

**Comment by pablosichert** [comment]
This PR will be merged once https://github.com/vectordotdev/vector/pull/10253 lands on master.

File: src/codecs/format/json.rs [hunk 3]
```diff
@@ -80,13 +84,60 @@ impl From<&JsonDeserializerConfig> for JsonDeserializer {
...
+impl Encoder<Event> for JsonSerializer {
+    type Error = crate::Error;
+
+    fn encode(&mut self, event: Event, buffer: &mut bytes::BytesMut) -> Result<(), Self::Error> {
+        let json = match event {
+            Event::Log(log) => serde_json::to_vec(&log),
+            Event::Metric(metric) => serde_json::to_vec(&metric),
+        }?;
+
+        buffer.put(json.as_slice());
```
Comment (lukesteensen) [inline src/codecs/format/json.rs:125]: Probably better to use `to_writer` and avoid the extra allocation + copy here.

Comment (pablosichert) [reply src/codecs/format/json.rs]: Didn't know about that API, very neat!

Comment (pablosichert) [reply src/codecs/format/json.rs]: Applied that suggestion in 650c17f4e61710b942b574687e491a6410dd5f88.

**Pull request merged**